### PR TITLE
Allows passing metrics key name in sharedmain.Main()

### PR DESCRIPTION
**Pull request by hrishin**
The sharedmain.Main() accept only one key "component name" as logging key and it uses the
same key name as a prefix for emitting the metrics on /metrics endpoint. In the case of Tekton
we want to use a different name for the component and different prefix of metrics name.
sharedmain.Main could accept the additional argument to pass metrics prefix name.

This patch enhances sharedmain.Main() to accept additional parameter
as a metrics name.

Fixes https://github.com/knative/pkg/issues/779

### CODE REVIEW

File: injection/sharedmain/main.go [hunk 1]
```diff
@@ -90,12 +90,12 @@ func GetLoggingConfig(ctx context.Context) (*logging.Config, error) {
 	return logging.NewConfigFromConfigMap(loggingConfigMap)
 }
 
-func Main(component string, ctors ...injection.ControllerConstructor) {
+func Main(component, metricsName string, ctors ...injection.ControllerConstructor) {
```
Comment (vagababov) [inline injection/sharedmain/main.go]: This is a backwards incompatible change.
What you probably should do instead is create a new `MainWithMetricName` and thread that through there.

Comment (hrishin) [reply injection/sharedmain/main.go]: @vagababov yes you are right.

**Comment by hrishin** [comment]
@vagababov thats valid question :thinking:  I thought `MainWithConfig` by by notion is a kind of auxiliary function (I guess its not).
However, it's better to keep it backward compatible as well (though increasing nesting may not look good in the longterm?).

**Comment by vagababov** [comment]
Well we can just check eventing to see if they use it. I don't think serving does (but we can check as well).
If they don't, then we can just make it private, and stop adding more methods :)

**Comment by hrishin** [comment]
@vagababov it's used in [plenty of places](https://github.com/search?q=org%3Aknative+MainWithConfig&type=Code) such as operator and controllers :sweat_smile:

**Comment by vagababov** [comment]
Well I guess the alternative is to stage the change in all the repos and then do a fast one, since imploding number of constructors is not very useful :

**Comment by hrishin** [comment]
> Well I guess the alternative is to stage the change in all the repos and then do a fast one

Do you mean to raise the PR ready? Could you elaborate on what exactly this means? :thinking: 

How about getting this patch in while marking unwanted constructor as deprecating? Then eventually migrate all other dependencies and deprecate those constructors?

**Comment by vagababov** [comment]
I mean we can create pending PRs in all repos and then fast merge them all
in sequence.
Your suggestion might also might work.

**Comment by hrishin** [comment]
> I mean we can create pending PRs in all repos and then fast merge them all
in sequence.

Alright @vagababov will get that work

**Comment by mattmoor** [comment]
I worry about the ever-expanding arity of constructor overloads, especially when they take N strings.  Perhaps we could provide a way to attach a metric name to context and default it to the component name if omitted.  That way we can simply fallback on the variant that allows a context to be provided?

**Comment by n3wscott** [comment]
+1 to main with context
-1 to adding to the signature.

**Comment by hrishin** [comment]
@mattmoor @n3wscott makes sense. (I was away, procastinating)

Will try to get this in. :+1: Thank you for being patient here. :bowing_man:

**Comment by vagababov** [comment]
Stale with merge conflicts and no activity for 5 months.
/close

**Pull request closed without merging**